### PR TITLE
[Telemetry Operator] Secrets watcher

### PR DESCRIPTION
**Pull request by rakesh-garimella**
**Description**

Changes proposed in this pull request:

- Add a mechanism that caches secrets referenced by all log pipelines. Also, watch all secret updates in the cluster - if the secret is cached, then trigger a reconciliation of the corresponding log pipelines
- Add an e2e test
- Decrease unnecessary log lines to the debug log level

**Related issue(s)**

### CODE REVIEW

File: components/telemetry-operator/controller/logpipeline/controller.go [hunk 1]
```diff
@@ -81,9 +84,42 @@ func NewReconciler(
 func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&telemetryv1alpha1.LogPipeline{}).
+		Watches(
+			&source.Kind{Type: &corev1.Secret{}},
+			handler.EnqueueRequestsFromMapFunc(r.enqueueRequests),
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(event event.CreateEvent) bool {
```
Comment (rakesh-garimella) [inline components/telemetry-operator/controller/logpipeline/controller.go:91]: Predicate function is pretty intrusive, please check if with this change we can still create a new pipeline or not. In my tests I had some issues that creating a new pipeline was rejected. So I had to do type assertion to verify if incoming object is a secret

Also this change will not cover the case where a user creates a pipeline which is referencing a secret that does not exist yet and would be create later.

Comment (lindnerby) [reply components/telemetry-operator/controller/logpipeline/controller.go:91]: I think you were using the filter by predicates on the eventhandler registration level in `WithEventFilter` not on the `Watches` registration where events are already filtered by type Secret and then passed on to the predicates function. You are correct, it wasn't working with that setup so we needed to change it.
We tested it thoroughly with the current changes and they seem fine.

File: components/telemetry-operator/apis/telemetry/v1alpha1/logpipeline_types.go [hunk 2]
```diff
@@ -273,6 +278,15 @@ func (l *LogPipeline) ContainsCustomPlugin() bool {
 	return l.Spec.Output.IsCustomDefined()
 }
 
+func ContainsCustomPlugin(l *LogPipeline) bool {
```
Comment (chrkl) [inline components/telemetry-operator/apis/telemetry/v1alpha1/logpipeline_types.go:281]: Do we need this?

File: components/telemetry-operator/controller/logpipeline/controller.go [hunk 3]
```diff
@@ -252,13 +307,13 @@ func (r *Reconciler) updateMetrics(allPipelines *telemetryv1alpha1.LogPipelineLi
 type keepFunc func(*telemetryv1alpha1.LogPipeline) bool
 
 func count(pipelines *telemetryv1alpha1.LogPipelineList, keep keepFunc) int {
-	count := 0
+	c := 0
```
Comment (chrkl) [inline components/telemetry-operator/controller/logpipeline/controller.go:310]: Was this change intentional?

File: components/telemetry-operator/controller/logpipeline/controller.go [hunk 4]
```diff
@@ -131,16 +171,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	}
 
 	if changed {
-		log.V(1).Info("Fluent Bit configuration was updated. Restarting the DaemonSet")
+		log.Info("Fluent Bit configuration was updated. Restarting the DaemonSet")
```
Comment (chrkl) [inline components/telemetry-operator/controller/logpipeline/controller.go:174]: This PR reaplces `log.Info` by `log.V(1).Info` for most of the occurrences (I guess for consistency?). Is there any reason to do it vice-versa in this line?

Comment (skhalash) [reply components/telemetry-operator/controller/logpipeline/controller.go:174]: Yeah, most of the info logs created too much noise, so they were demoted to debug. However, this one is good to have with the info level in order to see which reconciliation actually triggered a restart.

File: components/telemetry-operator/controller/logpipeline/controller.go (hunk 3 above)
Comment (skhalash) [reply components/telemetry-operator/controller/logpipeline/controller.go:310]: I assume, GoLand was complaining since the variable name shadowed the func name, which was also count.

**Review by chrkl — APPROVED** [review]
